SmartDeviceLink/SDLStreamingMediaLifecycleManager.m

```diff
@@ -0,0 +1,544 @@
+    [self sdl_sendBackgroundFrames];
+    [self.appStateMachine transitionToState:SDLAppStateInactive];
+}
+
+- (void)didEnterStateIsRegainingActive { }
+
+#pragma mark Video Streaming
++ (NSDictionary<SDLState *, SDLAllowableStateTransitions *> *)sdl_videoStreamStateTransitionDictionary {
+    return @{
+             SDLVideoStreamStateStopped : @[SDLVideoStreamStateStarting],
+             SDLVideoStreamStateStarting : @[SDLVideoStreamStateStopped, SDLVideoStreamStateReady],
+             SDLVideoStreamStateReady : @[SDLVideoStreamStateShuttingDown, SDLVideoStreamStateStopped],
+             SDLVideoStreamStateShuttingDown : @[SDLVideoStreamStateStopped]
+             };
+}
+
+- (void)didEnterStateVideoStreamStopped {
+    _videoEncrypted = NO;
+    
+    if (_videoEncoder != nil) {
+        [_videoEncoder stop];
+        _videoEncoder = nil;
+    }
+    
+    [[NSNotificationCenter defaultCenter] postNotificationName:SDLVideoStreamDidStopNotification object:nil];
+    
+    if (self.shouldRestartVideoStream) {
+        self.restartVideoStream = NO;
+        [self sdl_startVideoSession];
+    }
+}
+
+- (void)didEnterStateVideoStreamStarting {
+    self.restartVideoStream = NO;
+    
+    if (self.requestedEncryptionType != SDLStreamingEncryptionFlagNone) {
+        [self.protocol startSecureServiceWithType:SDLServiceTypeVideo completionHandler:^(BOOL success, NSError *error) {
+            // This only fires if we fail
+            if (error) {
+                SDLLogE(@"TLS setup error: %@", error);
+                [self.videoStreamStateMachine transitionToState:SDLVideoStreamStateStopped];
+            }
+        }];
+    } else {
+        [self.protocol startServiceWithType:SDLServiceTypeVideo];
+    }
+}
+
+- (void)didEnterStateVideoStreamReady {
+    if (_videoEncoder == nil) {
+        NSError* error = nil;
+        _videoEncoder = [[SDLVideoEncoder alloc] initWithDimensions:self.screenSize properties:self.videoEncoderSettings delegate:self error:&error];
+        
+        if (error) {
+            SDLLogE(@"Could not create a video encoder: %@", error);
+            [self.videoStreamStateMachine transitionToState:SDLVideoStreamStateStopped];
+            return;
+        }
+        
+        if (!self.backgroundingPixelBuffer) {
+            CVPixelBufferRef backgroundingPixelBuffer = _videoEncoder.pixelBuffer;
+            if (CVPixelBufferAddText(backgroundingPixelBuffer, @"") == NO) {
+                SDLLogE(@"Could not create a backgrounding frame");
+                [self.videoStreamStateMachine transitionToState:SDLVideoStreamStateStopped];
+                return;
+            }
+            
+            self.backgroundingPixelBuffer = backgroundingPixelBuffer;
+        }
+    }
+    
+    [[NSNotificationCenter defaultCenter] postNotificationName:SDLVideoStreamDidStartNotification object:nil];
+}
+
+- (void)didEnterStateVideoStreamShuttingDown {
+    [self.protocol endServiceWithType:SDLServiceTypeVideo];
```

Comment:
Video encoder gets stopped once the state heads to Stopped. This only occurs if there's an error or if we receive a stop session ACK.